### ISSUE TITLE
List env vars with descriptions

### ISSUE THREAD

**Issue by marcingrzejszczak** [zdharma/zshelldoc#6]
Hi!

Once again, great work with the framework! :)

I was thinking about the env vars. Currently, they are printed in a nice way as a list after each function. Maybe it could be useful to somehow provide all environment variables in a table. Also, we could provide env vars with a description of some sort. What do you think?

**Comment by psprint** [comment]
It might be indeed better to use a table than space-saving enumeration with quite custom separator (the "//"). The descriptions can be problematic, single line can export multiple env vars, hard to decide from where should the description come from. Maybe a special comment in script-body, like:

```zsh
# env-vars {{{
# findLatestProdTag -> LATEST_PROD_TAG -> Here is the description text
# parsePipelineDescriptor -> PIPELINE_DESCRIPTOR_PRESENT -> Here is the description text, new-lines are problematic
# }}}
```

Just doing a brain storm. It might be that the table and descriptions would be better as a standalone section in the document.

That said I've received a feature-request with donation for my other project Zplugin. It's a little big thing, I'll be busy for a week, but maybe we can work out good TODO during that time.

**Comment by psprint** [comment]
The format of the section looks promising. Is it OK to assume that the env-vars section will be in script body, not in a function? It was natural for synopsis section, env-vars could theoretically be located in a function that exports everything (the L515-518 link to source has bitrot so unsure if that was what you wanted, could you edit, best via entering commit and "Browse files"). Quite simple implementation would allow this section in function comment.

So the descriptions should be in each function, that's fine. I'm just not comfortable with making function sections (in the resulting document) longer, but that's probably just wrong point of view, wanting to make function sections space-sparing, raw, short structures (enumerations, etc.), probably making such sections harder on eyes. So adding some more free text can be a good idea.

Could a donation be made for this, like $10? Sorry that I'm asking, I'm just tired with 3 active projects and using this money to refresh myself and stay active.

**Comment by marcingrzejszczak** [comment]
Yeah, definitely we can have the descriptions in the script body.


As for the donation, this library is used by another open source project. I can definitely wait until you have some spare time. There's no rush

**Comment by psprint** [comment]
Ok I let you know when I code something. Can I know the project name?

**Comment by marcingrzejszczak** [comment]
Of course, I think I've already told you that. It' Spring Cloud Pipelines https://github.com/spring-cloud/spring-cloud-pipelines/ and its new fork https://github.com/CloudPipelines/scripts

**Comment by psprint** [comment]
I've approached this. The most difficult part – parsing `env-vars {{{ A -> B\nC->D }}}` block is done, and a table is being output:

![env-vars](https://user-images.githubusercontent.com/6049288/45894619-30d5d480-bdcf-11e8-85e9-b36bf533ca8a.png)

Our initial idea was to provide the descriptions inside using (or exporting, hm) functions. This can still be done. I think that having table with all variables at top of document is a good idea even with per-function descriptions.

Do you think this per-function stuff could be reduced into: make environment variables  anchors (in the `... // ... // ...` enumerations) pointing to the top table? If not then I'll code per-function tables with descriptions.

I've used following env-vars block within script body (under synopsis) of `pipeline.sh`:

```
# env-vars {{{
# LATEST_PROD_TAG -> latest product
#                    tag
#   PASSED_LATEST_PROD_TAG   ->   accepted latest product tag
# PIPELINE_DESCRIPTOR_PRESENT -> marks if
#                                the pipeline descriptor
# is present
# PARSED_YAML -> YAML contents fully processed
# }}}
```

I would appreciate an $3 donation if that's possible, I've put updated PayPal/Patreon info on the top of README.

**Comment by marcingrzejszczak** [comment]
Thanks for the update. Like I said:

> this library is used by another open source project. I can definitely wait until you have some spare time. There's no rush

**Comment by psprint** [comment]
Ok. Do you think the documentation of an env-var should be included in section devoted in function that's using or exporting it? I guess that's the "want everything" thing, not a "hesitate to provide" attitude, so it probably should be done this way?

**Comment by marcingrzejszczak** [comment]
Hmm interesting question. What if you have multiple functions using the same env var? We would have a duplicated environment variable description. OTOH you won't have to switch back and forth. Another option is to have a single table with env vars and their descriptions and in the function just a link to that table. WDYT?
